workflows-cli/src/command_runner.rs

```diff
@@ -0,0 +1,143 @@
+use serde::Deserialize;
+use std::collections::HashMap;
+use std::env;
+use std::fs;
+use std::io;
+use std::os::unix::process::ExitStatusExt;
+use std::process::{Command, Output};
+
+/// Underlying traits used to run/mock CLI commands
+pub trait CommandLike {
+    /// Command arguments
+    fn arg(&mut self, arg: &str) -> &mut dyn CommandLike;
+    /// Execute command
+    fn output(&mut self) -> io::Result<Output>;
+}
+
+/// Real command runner
+pub struct RealCommand {
+    /// Command instance
+    cmd: Command,
+}
+
+/// Add arg and ouput methods to Realcommand
+impl CommandLike for RealCommand {
+    fn arg(&mut self, arg: &str) -> &mut dyn CommandLike {
+        self.cmd.arg(arg);
+        self
+    }
+
+    fn output(&mut self) -> io::Result<Output> {
+        self.cmd.output()
+    }
+}
+
+/// Shared 'new' command trait between real and mock runners
+pub trait CommandFactory {
+    /// Returns a CommandLike runner
+    fn new_command(&self, cmd: &str) -> Box<dyn CommandLike>;
+}
+
+/// Return a real command runner
+pub struct RealCommandFactory;
+
+/// Heap allocated command to assign command runner at run-time
+impl CommandFactory for RealCommandFactory {
+    fn new_command(&self, cmd: &str) -> Box<dyn CommandLike> {
+        Box::new(RealCommand {
+            cmd: Command::new(cmd),
+        })
+    }
+}
+
+/// Structure of mock entry as in mapping file
+#[derive(Debug, Deserialize, Default, Clone)]
+struct MockEntry {
+    /// Calling command
+    command: String,
+    /// CLI return
+    response: String,
+    /// Exit code
+    code: i32,
+}
+
+/// Required fields to mock real command
+pub struct MockCommand {
+    /// Command being used
+    cmd: String,
+    /// CLI arguents
+    args: Vec<String>,
+    /// Map CLI call to output
+    mappings: Vec<MockEntry>,
+}
+
+impl MockCommand {
+    /// Initiate mock command runner
+    pub fn new(cmd: &str) -> Self {
+        let yaml_str = fs::read_to_string("./tests/mock_commands.yaml")
+            .expect("Failed to read mock_commands.yaml");
+
+        let mappings: HashMap<String, Vec<MockEntry>> =
+            serde_yaml::from_str(&yaml_str).expect("Failed to parse YAML");
+
+        let target_mapping = env::var("RUST_TEST_ACTIVE_MAPPING").unwrap();
```

Comment:
It would be best to "namespace" any special environment variables that our application depends on.

eg WORKFLOWS_CLI_TEST_ACTIVE_MAPPING or similar